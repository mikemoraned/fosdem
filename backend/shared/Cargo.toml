[package]
name = "shared"
version = "0.0.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
openai_dive = { workspace = true }
sqlx = { workspace = true }
pgvector = { workspace = true }
log = { workspace = true }
url = { workspace = true }
