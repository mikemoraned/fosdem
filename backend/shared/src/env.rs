use tracing::info;

pub fn load_secret(name: &str) -> Result<String, Box<dyn std::error::Error>> {
    let secret = dotenvy::var(name).map_err(|e| format!("{} is not set, e: {:?}", &name, e))?;
    let suffix = secret[(secret.len() - 3)..].to_string();
    info!(
        "Loaded secret with name '{}', ending with '{}'",
        name, suffix
    );
    Ok(secret)
}

pub fn load_public(name: &str) -> Result<String, Box<dyn std::error::Error>> {
    let value = dotenvy::var(name).map_err(|e| format!("{} is not set, e: {:?}", &name, e))?;
    info!("Loaded public item with name '{}', value '{}'", name, value);
    Ok(value)
}
