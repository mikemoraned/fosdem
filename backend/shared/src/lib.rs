pub mod cli;
pub mod env;
pub mod openai;
pub mod postgres_openai;
pub mod queryable;
