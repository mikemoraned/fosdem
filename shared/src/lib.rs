pub mod cli;
pub mod env;
pub mod inmemory_openai;
pub mod model;
pub mod openai;
pub mod queryable;
