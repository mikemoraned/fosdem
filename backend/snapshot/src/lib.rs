use serde::Serialize;
use shared::{inmemory_openai::InMemoryOpenAIQueryable, model::SearchItem, queryable::Queryable};

#[derive(Debug, PartialEq, Serialize)]
pub struct DistanceSummary {
    distance: f64,
    event_id: u32,
    event_title: String,
}

pub struct Snapshotter {
    queryable: InMemoryOpenAIQueryable,
}

impl Snapshotter {
    pub async fn new(
        openai_api_key: &str,
        model_dir: &std::path::Path,
    ) -> Result<Snapshotter, Box<dyn std::error::Error>> {
        let queryable = InMemoryOpenAIQueryable::connect(model_dir, openai_api_key)
            .await
            .unwrap();
        Ok(Snapshotter { queryable })
    }

    pub async fn search(
        &self,
        title: &str,
    ) -> Result<Vec<DistanceSummary>, Box<dyn std::error::Error>> {
        Ok(Snapshotter::summarise(
            &self.queryable.search(title, 20, true).await.unwrap(),
        ))
    }

    fn summarise(items: &[SearchItem]) -> Vec<DistanceSummary> {
        items
            .iter()
            .map(|i| DistanceSummary {
                distance: i.distance,
                event_id: i.event.id,
                event_title: i.event.title.clone(),
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use std::{path::PathBuf, str::FromStr};

    use shared::env::load_secret;

    use super::*;

    static PHRASES: [&str; 1] = ["controversial"];

    #[tokio::test]
    async fn test_phrase_search() {
        let openai_api_key = load_secret("OPENAI_API_KEY").unwrap();
        let model_dir = PathBuf::from_str("../shared/data/model").unwrap();
        let snapshotter = Snapshotter::new(&openai_api_key, &model_dir).await.unwrap();
        for phrase in PHRASES {
            let similar = snapshotter.search(phrase).await.unwrap();
            insta::assert_yaml_snapshot!(similar);
        }
    }
}
