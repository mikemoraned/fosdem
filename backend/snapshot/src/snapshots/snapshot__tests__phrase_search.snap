---
source: snapshot/src/lib.rs
expression: similar
---
- distance: 0.6884902032907187
  event_id: 2405
  event_title: Best practices for research in open source ecosystems
- distance: 0.6945371774062765
  event_id: 2779
  event_title: "Staying Ahead of the Game: JavaScript Security"
- distance: 0.6971099789086412
  event_id: 1930
  event_title: OpenPrinting - We make printing just work!
- distance: 0.697975287430046
  event_id: 3134
  event_title: Another approach to AI
- distance: 0.7006308115511992
  event_id: 2747
  event_title: "For Want of Anneal: Examining The Unseen Changes Concerning Changes To VCS Assets and The Need For More Graph Centric Approaches"
- distance: 0.7019730046401185
  event_id: 2866
  event_title: "Making API Terms of Service Trustworthy? Presenting FACT : The Fair API Commitment Terms"
- distance: 0.7035085263300751
  event_id: 3553
  event_title: Learning from disaster response teams to save the internet
- distance: 0.7043562493209279
  event_id: 3668
  event_title: "Hot Topics: Organizers of the Legal & Policy DevRoom discuss the issues of the day"
- distance: 0.7050657161828311
  event_id: 2638
  event_title: "Breaking Barriers: Content Management Systems and Accessibility"
- distance: 0.7055400049860489
  event_id: 3544
  event_title: "Defend FOSS: From innovation to world-wide positive change"
- distance: 0.7069287942388786
  event_id: 3163
  event_title: "Copyleft and the GPL: Finding the Path Forward to Defend our Software Right to Repair"
- distance: 0.7077077121574527
  event_id: 3113
  event_title: RHEL and CentOS and the growth of openwashing in FOSS
- distance: 0.7084080793696588
  event_id: 3717
  event_title: "FOSS policy engagement: panel"
- distance: 0.7086130969402366
  event_id: 2750
  event_title: "Codes Bound by Ethics: The Rising Tide of Non-Free Software Licenses in AI ecosystems"
- distance: 0.710486307096886
  event_id: 2079
  event_title: "Fortify AI against regulation, litigation and lobotomies"
- distance: 0.7106596488330573
  event_id: 2310
  event_title: Meritocracy or Do-ocracy - why diversity is still hard and what can we do
- distance: 0.7108114986401424
  event_id: 1953
  event_title: "Using code generated by AI: issues, misconceptions and solutions"
- distance: 0.7110709188763001
  event_id: 3364
  event_title: Why not run OpenCL-accelerated LLM on your phone?
- distance: 0.7113728605409577
  event_id: 3234
  event_title: A journey accross the environmental materiality of digital services
- distance: 0.7118929963742723
  event_id: 3704
  event_title: "Fireside Chat on Further Restrictions, Imposed Downstream on Copyleft, Wreaking Havoc"
